Support CRSF as a bi-directional output protocol

andyp1per opened the pull request:
This allows CRSF to act in a similar way to SBUS Out, but bi-directional, full-duplex at 2Mbaud.

The primary use-case is dual flight controller setups where one flight controller is the "rate controller" (i.e. operating at speed in real-time) and the other flight controller is the "EKF / position controller" (i.e. can operate with fewer real-time constraints). It allows both of these controllers to be the same software stack or different stacks. So I have proved that an ardupilot<->ardupilot setup works and an ardupilot<->betaflight setup works.

Use cases:
- clean multi-core implementations (e.g. one core running rate controller the other running the EKF controller)
- heterogenous CPU implementations, e.g. rate controller on M7 (better realtime but less compute/memory/flash) EKF controller on A9 (worse realtime but better compute/memory/flash)
- heterogenous FC implementations where ArduPilot is not supported on the full stack (e.g. RP3250B for betaflight, PiZero2W for ArduPilot)
- Testing - provides full end-to-end testing of the CRSF stack which we did not have before.

This can be used to drive one flight controller from another flight controller using existing rate output support. 

The new protocol is 51 and so can be used by setting:

```SERIALx_PROTOCOL = 51```

This PR has been tested on Raspberry Pi's new drone, codename "Wolverine" and the work has been kindly sponsored by Raspberry Pi. 

The follow-up PR relies on https://github.com/betaflight/betaflight/pull/14870 which has now been merged

andyp1per committed:
AP_SerialManager: add CRSF output protocol

andyp1per committed:
AP_OSD: add CRSF output protocol

andyp1per committed:
AP_Vehicle: add update_rc() so that slow updates can be performed by the rc library

andyp1per committed:
AP_Math: create crc8_dvb_update_generic() for use by CRSF in general

andyp1per committed:
AP_CRSF: refactor key data structures and parsing functionality of CRSF into AP_CRSF_Protocol

refactor common frame encoding into AP_CRSF_Protocol
enable AP_CRSF_Protocol to be a virtual base class

andyp1per committed:
AP_CRSF: implementation of CRSF output protocol

correctly output servo channels in CRSF out channels
negotiate CRSFv3 at 2Mbaud for CRSF out channels
add support for ping frames and receiving vesion info on CRSF out channels
support device pings to give the remote end correct version information
CRSF output channels manage their own frame encoding
propagate fake link statistics tx frames to the remote end when using CRSF out
refactor out management into a separate class
normalize CRSF output thread to regular beats and only send rc frames at 250Hz

andyp1per committed:
AP_RCProtocol: add support for protocol detection debug

andyp1per committed:
AP_RCTelemetry: move key data structures and defines into AP_CRSF_Protocol

refactor the definiton of CommandFrame to AP_CRSF_Protocol
refector version parsing code into AP_CRSF_Protocol
rename use of DeviceAddress
refactor enums into AP_CRSF_Protocol

andyp1per committed:
AP_RCTelemetry: add support for CRSF output channels

refactor to support working CRSF output bootstrap

timtuxworth commented:
WOW!

andyp1per committed:
hwdef: don't include CRSF output in Pixhawk1-1M

andyp1per committed:
AP_RCProtocol: support Betaflight as a peer for CRSF

peterbarker commented on ArduCopter/Copter.h:
```diff
@@ -607,7 +607,8 @@ class Copter : public AP_Vehicle {
         AUTO_DO_LAND_START = 6,
-        BRAKE_LAND         = 7
+        BRAKE_LAND         = 7,
+        DISARM             = 8
```
```suggestion
        DISARM             = 8,
```

peterbarker commented on ArduCopter/events.cpp:
```diff
@@ -514,6 +517,9 @@ void Copter::do_failsafe_action(FailsafeAction action, ModeReason reason){
             set_mode_brake_or_land_with_pause(reason);
             break;
+        case FailsafeAction::DISARM:
+            arming.disarm(AP_Arming::Method::RADIOFAILSAFE);
```
Don't you really want to pass "false" here for disarm checks?

peterbarker commented on libraries/AP_RCProtocol/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,388 @@
+ * from the main SRV_Channels and telling its underlying CRSF protocol instance
+ * to send them at a user-configurable rate.
+ */
+#include <AP_RCProtocol/AP_RCProtocol_config.h>
```
Not sure if this belongs in here or like AP_SBusOut as a separate library.

We only do input in this directory ATM.

peterbarker commented on libraries/AP_RCProtocol/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,388 @@
+AP_CRSF_Out* AP_CRSF_Out::_singleton;
+
+AP_CRSF_Out::AP_CRSF_Out(void) :
+    _state(State::WAITING_FOR_PORT)
```
Put this initialisation in the header where the declaration of the variable is made.

peterbarker commented on libraries/AP_RCProtocol/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,388 @@
+        return;
+    }
+
+    _crsf_port = AP_RCProtocol_CRSF::get_direct_attach_singleton(AP_SerialManager::SerialProtocol_CRSF_Output, 0);
```
I can see this might be one of the reasons you put this in this directory.  Presumably we really should be calling AP_CRSF_Telem to get this object if we were to put this in a different directory.

peterbarker commented on libraries/AP_RCProtocol/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,388 @@
+    _state = State::WAITING_FOR_RC_LOCK;
+}
+
+bool AP_CRSF_Out::do_status_update()
```
```suggestion
bool AP_CRSF_Out::should_do_status_update()
```

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_RCProtocol/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,388 @@
+    const uint8_t nchan = MIN(NUM_SERVO_CHANNELS, (uint8_t)CRSF_MAX_CHANNELS);
+
+    for (uint8_t i = 0; i < nchan; ++i) {
+        SRV_Channel *c = SRV_Channels::srv_channel(i);
```
So this means you're going to get everything post-mixing.  Is that the desired outcome?  This is directly analogous to the (non-mixing) IOMCU firmware or S.Bus output.

I thought your intent here was to have something else do the rate loops - I can't see how this does that without very specific servo output settings.

peterbarker replied:
> Yeah this is correct. The target system wants to get the AP modified channels including k_roll_out and friends. Trust me it works :)

Gah!  I was a little misled by my first reading here.
My question was, "get everything post-mixing?" and you replied "yeah".   But that's not actually the case - if you are using k_roll_out and friends it's pre-motor-mixing.
This means:

this entire feature only works if you are using AP_MotorsMulticopter
throttle has been filtered (including slewing)
but I don't think it has been linearised

... come to think of it.... why are you going via servo outputs at all?!  Why not have AP_MotorsOutput_CRSF_AETR?  It would be a better structure *and* more efficient.  This file would either become AP_MotorsOutput_CRSF_AETR or this file would read state from AP_MotorsOutput_CRSF_AETR or AP_MotorsOutput_CRSF_AETR would write state in here.  But notionally you could send the AETR packet as soon as the motor library updated its output.

peterbarker replied:
We discussed this on DevCall.

Going via SRC_Channels notionally means that a Plane can use this for - supplying rate outputs in the same manner (i.e. via SRV_Channels).  This bears thinking about.

peterbarker commented on libraries/AP_RCProtocol/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,388 @@
+        debug_rcout("Updating channels. CH1=%u CH2=%u CH3=%u", channels[0], channels[1], channels[2]);
+    }
+#endif
+    AP_CRSF_Protocol::Frame frame;
```
Either zero this or (more correctly declare it `const` and fill in all of the fields with initialisers).

In either case, don't make it looks like we might use uninitialised stack data.

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -744,90 +729,55 @@ void AP_CRSF_Telem::process_vtx_telem_frame(VTXTelemetryFrame* vtx)
 {
     debug("process_command_frame: 0x%x -> 0x%x: 0x%x", command->origin, command->destination, command->payload[0]);
-    if (command->destination != 0 && command->destination != AP_RCProtocol_CRSF::CRSF_ADDRESS_FLIGHT_CONTROLLER) {
+    if (command->destination != 0 && command->destination != DeviceAddress::CRSF_ADDRESS_FLIGHT_CONTROLLER) {
```
```suggestion
    if (command->destination != 0 && command->destination != DeviceAddress::FLIGHT_CONTROLLER) {
```

(and these should be `enum class`)

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -744,90 +729,55 @@ void AP_CRSF_Telem::process_vtx_telem_frame(VTXTelemetryFrame* vtx)
 
     switch (command->payload[0]) {
-        case AP_RCProtocol_CRSF::CRSF_COMMAND_GENERAL_CRSF_SPEED_PROPOSAL: {
+        case CommandGeneral::CRSF_COMMAND_GENERAL_CRSF_SPEED_PROPOSAL: {
```
```suggestion
        case CommandGeneral::SPEED_PROPOSAL: {
```

etc

peterbarker reviewed (changes requested):
I think my main comment here is that CRSF-out definitely doesn't belong anywhere near rc input.

It should be near SBusOut as much as possible if it is directly analogous.

The clocking might mean that we can't put the call to emit outputs in SRC_Channels - doing so would potentially reduce latency, mind you.

Partial review only

andyp1per committed:
AP_Vehicle: add CRSF output management

compile out crsf out scheduled task when neccessary

andyp1per committed:
AP_RCProtocol: normalize CRSF output thread to regular beats and only send rc frames at 250Hz

andyp1per committed:
AP_RCProtocol: refactor to support direct-attach CRSF output channels

add CRSF out to update calls
allow CRSF uart to be started externally and use unbuffered writes for speed

andyp1per committed:
waf: add AP_CRSF module

andyp1per committed:
AP_Periph: add AP_CRSF module

andyp1per committed:
AP_RCProtocol: telem_available does not need to be mutable

andyp1per committed:
AP_IOMCU: add AP_CRSF to build

andyp1per committed:
AP_RCProtocol: remove crsf out config

address config of CRSF

andyp1per committed:
autotest: update PerfInfo test for updated vehicle scheduler table

peterbarker commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,437 @@
+{
+    if (_singleton == nullptr) {
+        _singleton = this;
+    }
```
This is weird - you can only have one crsf out backend?

peterbarker commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,437 @@
+        return false;
+    }
+
+    _crsf_port = NEW_NOTHROW AP_RCProtocol_CRSF(AP::RC(), AP_RCProtocol_CRSF::PortMode::DIRECT_RCOUT, uart);
```
There should be a common base class for this and AP_RCProtocol_CRSF - we shouldn't be creating an RC protocol backend here.

peterbarker commented on libraries/AP_HAL_ChibiOS/hwdef/include/minimize_features.inc:
```diff
@@ -38,4 +38,6 @@ define AP_MOTORS_FRAME_HEXA_ENABLED 1
\ No newline at end of file
+define AP_ICENGINE_ENABLED 0
+
+define AP_CRSF_OUT_ENABLED 0
```
This should be off by default on all vehicles, not just minimized vehicles (IMO).

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.h:
```diff
@@ -503,3 +481,4 @@ namespace AP {
 };
 
 #endif
+
```
stray whitespace

andyp1per replied:
thanks fixed

peterbarker reviewed (changes requested):
partial review only

andyp1per commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,437 @@
+
+        // Calculate time remaining until the next scheduled frame to maintain precise rate
+        uint32_t next_run = _last_frame_us + _frame_interval_us;
+        uint8_t frame_ratio = _heartbeat_to_frame_ratio;
```
No?

andyp1per committed:
AP_CRSF: introduce specific config item for CRSF

correct timeout calculations in output rates
only allow a single CRSF out instance
optimise CRSF protocol in compilation
add vanity comments and address review comments

andyp1per commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,437 @@
+        // if we have not negotiated a faster baudrate do not go above the default output rate
+        if (uint16_t(_frontend._rate_hz.get()) > DEFAULT_CRSF_OUTPUT_RATE && _uart->get_baud_rate() == CRSF_BAUDRATE) {
+            next_run = _last_frame_us + 1000000UL / DEFAULT_CRSF_OUTPUT_RATE;
+            frame_ratio = 1;
```
Yes

andyp1per commented on libraries/AP_Vehicle/AP_Vehicle.cpp:
```diff
@@ -289,7 +290,13 @@ const AP_Param::GroupInfo AP_Vehicle::var_info[] = {
+    AP_SUBGROUPINFO(crsf_out, "CRSF_OUT_", 33, AP_Vehicle, AP_CRSF_OutManager),
+#endif
+
+AP_GROUPEND
```
thanks fixed

andyp1per commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,437 @@
+        const uint32_t now = AP_HAL::micros();
+
+        // Calculate time remaining until the next scheduled frame to maintain precise rate
+        uint32_t next_run = _last_frame_us + _frame_interval_us;
```
Thanks fixed

andyp1per committed:
scripts: add AP_CRSF_OUT_ENABLED to extract_features.py and build_options.py

andyp1per committed:
AP_HAL: disable CRSF output by default

andyp1per committed:
AP_RCProtocol: correct crsf output building for debug

andyp1per committed:
AP_Vehicle: initialise CRSF output early so that sensors are available for calibration

andyp1per commented on libraries/AP_CRSF/AP_CRSF_Out.h:
```diff
@@ -0,0 +1,135 @@
+    // low bandwidth heartbeat to provoke telemetry return
+    void send_heartbeat();
+
+    static AP_CRSF_Out* _singleton;
```
Fixed thanks

andyp1per committed:
AP_HAL: add tick-based minimal scheduler

add support for setting loop rate and task rate to minimal scheduler
add run_task_immediately()

andyp1per committed:
AP_CRSF: use minimal scheduler for sending frames

send AETR channels at loop rate and other channels at lower rate
attempt to run CRSF output rate at the same rate as the loop rate
add rate messages for CRSF output
send occasional ping frames to measure roundtrip latency
synchronize AETR channels to SRV_Channels::push()

tridge commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,443 @@
+    _instance_idx(instance), _uart(uart), _frontend(frontend)
+{
+    // in the future we could consider supporting multiple output handlers
+    // but since EAHRS is a singleton this may have limited utility
```
EAHRS?

peterbarker commented on libraries/AP_CRSF/AP_CRSF_OutManager.cpp:
```diff
@@ -0,0 +1,109 @@
+        AP_SerialManager::SerialProtocol protocol = port_state->get_protocol();
+        AP_HAL::UARTDriver* uart = serial_manager.get_serial_by_id(i);
+
+        if (_instances[i] != nullptr) {
```
move this sanity check up?

andyp1per replied:
Fixed

andyp1per committed:
AP_RCProtocol: move key data structures and defines to AP_CRSF_Protocol

refactor enums into AP_CRSF_Protocol
refactor common frame encoding into AP_CRSF_Protocol
refactor variable bit channel decoding into AP_CRSF_Protocol
support new link rate field for reporting link rate
move HeartbeatFrame into AP_CRSF_Protocol

andyp1per committed:
SRV_Channels: call AP_CRSF_Out::push() in push()

andyp1per committed:
autotest: two vehicle CRSF autotest

andyp1per committed:
AP_RCProtocol: drop HEARTBEAT frames in CRSF rather than just registering decoding as failed

andyp1per committed:
AP_RCProtocol: allow 2MB CRSF baud rate on SITL

andyp1per committed:
AP_CRSF: move SubsetChannelsFrame to AP_CRSF_Protocol and encode variable bit channels correctly

peterbarker reviewed (changes requested):
Is sending our servo outputs as RC inputs *really* the best way to actually do all of this?  It's an anti-pattern for most off-board control as often RC is scaled (e.g. expo) on the receiving side (among many other things).  And certainly our servo outputs are scaled on all sorts of different things.

```
Board                    AP_Periph  antennatracker  blimp  bootloader  copter  heli   iofirmware  plane  rover  sub
CubeOrange-periph-heavy  *                                 *                                                    
Durandal                            624             624    *           636     636                636    624    624
Hitec-Airspeed           *                                 *                                                    
KakuteH7-bdshot                     624             624    *           644     636                636    624    640
MatekF405                           560             560    *           572     572                580    568    560
Pixhawk1-1M-bdshot                  368             376                368     368                376    368    376
SITL_x86_64_linux_gnu               12568           16672              16704   12608              16704  12568  16672
f103-QiotekPeriph        *                                 *                                                    
f303-MatekGPS            0                                 *                                                    
f303-Universal           *                                 *                                                    
iomcu                                                                                 16                        
mindpx-v2                           560             552    *           572     572                572    560    560
revo-mini                           584             568    *           572     580                596    584    568
skyviper-v2450                                                         104                                      
speedybeef4                         576             568    *           572     580                596    576    568
```
What is this doing using up so much flash when the feature isn't enabled?!  This isn't just about the poor resource usage, it *also* means that there's extra code floating around for existing platforms making users *not* using the new feature at risk.

The number of boards this feature will be enabled on will be a rounding error on the number of boards the code will be flashed on.  The structure is a significant step backwards.

There's at least 4 ways you've used in here to pack packets.  Let's be consistent at least.

Shouldn't this be available on the custom build server?

I have previously suggested that this is the wrong way to develop.  I have suggested ways to do this which doesn't result in a PR which is problematic to review (and work on, I would imagine).

andyp1per committed:
AP_CRSF: address review comments

andyp1per committed:
AP_RCProtocol: change name of debug get_frame_type()

andyp1per committed:
AP_RCProtocol: address review comments

peterbarker commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,470 @@
+
+    scheduler.init(tasks, rate);
+    state = State::WAITING_FOR_RC_LOCK;
+    scheduler.set_task_rate(REPORTING, frontend.reporting_rate_hz);
```
Add enum name here

peterbarker commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,470 @@
+
+        // Check for overrun - if we are late by more than 50% of an interval,
+        // give up on the old timeline and reset.
+        if (AP_HAL::timeout_remaining(last_frame_us, now_us, interval_us + interval_us/2) == 0) {
```
should this be an elseif?  This can't happen if we've just delayed, correct?

peterbarker replied:
I didn't stop to think that the `delay_microseconds(timeout_remaining_us)` might itself cause the over-run.

peterbarker commented on libraries/AP_CRSF/AP_CRSF_Out.h:
```diff
@@ -0,0 +1,176 @@
+
+    TickScheduler _scheduler;
+
+    enum TaskIds {
```
Fair enough, but please do add the namespace wherever the thing is used.

peterbarker commented on libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp:
```diff
@@ -2259,7 +2196,7 @@ void AP_CRSF_Telem::get_multi_packet_passthrough_telem_data(uint8_t size)
   if is_tx_active is true then this will be a request for telemetry after receiving an RC frame
  */
-bool AP_CRSF_Telem::_get_telem_data(AP_RCProtocol_CRSF::Frame* data, bool is_tx_active)
+bool AP_CRSF_Telem::_get_telem_data(const AP_RCProtocol_CRSF* crsf_port, AP_CRSF_Protocol::Frame* data, bool is_tx_active)
```
crsf_port isn't used in here?

peterbarker reviewed (changes requested):
Several of my previous review comments have not been addressed

andyp1per committed:
AP_RCProtocol: minor build fix

andyp1per committed:
AP_RCTelemetry: minor build fix

davidbuzz reviewed (approved):
lets merge this. it passes tests and flies.

andyp1per committed:
AP_CRSF: address review comments

andyp1per committed:
AP_RCTelemetry: address review comments

andyp1per committed:
AP_RCProtocol: address review comments

andyp1per committed:
AP_RCProtocol: restore VTX standalone UART handling in RCIN instance

Move VTX (SerialProtocol_CRSF) port management back to the RCIN
AP_RCProtocol_CRSF instance, matching the original master behavior.
The RCIN constructor now finds and owns the standalone UART directly.

Remove the DIRECT_VTX port mode since standalone VTX is handled via
the _uart member on the PASSTHROUGH_RCIN instance. Remove VTX instance
creation from OutManager::init() as it is no longer needed there.

andyp1per committed:
AP_CRSF: make AP_CRSF_Out a singleton, remove OutManager

Convert AP_CRSF_Out to a proper singleton embedded directly in
AP_Vehicle. Move parameters (CRSF_OUT_RATE, CRSF_OUT_RPT_HZ) and
serial port discovery from the manager into AP_CRSF_Out itself.

The manager class added multi-instance infrastructure that was never
used (AP_CRSF_Out already enforced a single instance via panic).
Removing it saves flash and reduces complexity.

Delete AP_CRSF_OutManager.h and AP_CRSF_OutManager.cpp. Update
AP_Vehicle to embed AP_CRSF_Out directly and remove the 1Hz scheduled
init task since port discovery now happens once during startup.

andyp1per committed:
AP_CRSF: remove UARTDriver member from AP_CRSF_Out

The UART is already owned by the underlying AP_RCProtocol_CRSF
instance. Add a public get_baud_rate() accessor to AP_RCProtocol_CRSF
and use it from the CRSF output thread instead of keeping a separate
UART pointer. The UART is now a local in init() only.

andyp1per committed:
AP_CRSF: remove decode_11bit_channels

andyp1per committed:
AP_RCProtocol: reinstate old decode_11bit_channels being careful to match the size to the frame

andyp1per committed:
AP_CRSF: remove AP_CRSF_Protocol inheritance and dead update()

With the OutManager removed, the polymorphic update() dispatch
through AP_CRSF_Protocol has no callers. Remove the pure virtual
update() from AP_CRSF_Protocol, making it a pure utility class
(static methods and type definitions only). Remove inheritance
from AP_CRSF_Out and its dead empty update() implementation.

andyp1per committed:
AP_RCProtocol: remove AP_CRSF_Protocol multiple inheritance

AP_RCProtocol_CRSF no longer needs to inherit from AP_CRSF_Protocol
now that the pure virtual update() has been removed. The class already
inherits update() from AP_RCProtocol_Backend and accesses shared types
and static methods via AP_CRSF_Protocol:: qualification.

peterbarker commented on Tools/autotest/arducopter.py:
```diff
@@ -12680,8 +12680,144 @@ def PerfInfo(self):
+            self.wait_statustext("CRSFOut: RC: [0-9]+Hz,.*", regex=True, check_context=True, timeout=10)
+            self.progress("TX vehicle started CRSF output - connection established")
+
+            # Give time for negotiation
+            self.delay_sim_time(10)
```
Can we detect this rather than just delaying?

andyp1per replied:
I have added the *try to detect when we can move on* to my AI rules but I agree with Buzz these feel very arbitrary for a system that is totally undocumented.

peterbarker replied:
So what happens here when you do *not* delay?

What exactly are you waiting for?

peterbarker commented on Tools/autotest/arducopter.py:
```diff
@@ -12680,8 +12680,144 @@ def PerfInfo(self):
+            # On the RX side, verify we're receiving RC channels
+            self.progress("Checking RX vehicle receives RC channels")
+            # Put vehicle in stabilize so that there are actually some outputs on channel 1-4
+            self.change_mode('STABILIZE')
```
Why would there not be any outputs in the other modes?

peterbarker reviewed (commented):
.... why is this called AP_CRSF_Out when it also *receives* packets?

peterbarker reviewed (changes requested):
Sorry, that was a partial review only.  Hit the wrong button...

peterbarker reviewed (changes requested):
You know, I'm *still* not really sure what's being accomplished with all of this.

Could you explain in more detail what all of this achieves, please?

What capability does this bring to ArduPilot?



... this is looking much better now...

timtuxworth commented on Tools/autotest/arducopter.py:
```diff
@@ -12680,8 +12680,144 @@ def PerfInfo(self):
+                    "--serial5=tcp:%u" % crsf_port,  # listen for CRSF connection
+                ],
+                param_defaults={
+                    "SYSID_THISMAV": 2,
```
Should be MAV_SYSID

andyp1per commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,479 @@
+# define debug_rcout(fmt, args...)
+#endif
+
+#define DEFAULT_CRSF_OUTPUT_RATE      250U // equivalent to tracer
```
See I agree with this because someone could conceivably want to override this somewhere else in which case the pattern applies :)

andyp1per committed:
AP_CRSF: address review comments

andyp1per committed:
AP_RCProtocol: address review comments

andyp1per committed:
autotest: address review comments

andyp1per commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,479 @@
+
+bool AP_CRSF_Out::decode_crsf_packet(const AP_CRSF_Protocol::Frame& _frame)
+{
+#ifdef CRSF_RCOUT_DEBUG_FRAME
```
fixed

andyp1per commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,479 @@
+    frame.device_address = DeviceAddress::CRSF_ADDRESS_SYNC_BYTE;
+    frame.type = FrameType::CRSF_FRAMETYPE_SUBSET_RC_CHANNELS_PACKED;
+    uint8_t payload_len = AP_CRSF_Protocol::encode_variable_bit_channels(frame.payload, channels, nchan, start_chan);
+    frame.length = payload_len + 2; // +1 for type, +1 for CRC
```
but your persuaded me lower down ... so I have changed it

andyp1per commented:
> You know, I'm _still_ not really sure what's being accomplished with all of this.
> 
> Could you explain in more detail what all of this achieves, please?
> 
> What capability does this bring to ArduPilot?
> 
> ... this is looking much better now...

I added some text to the description

davidbuzz commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,479 @@
+        if (c != nullptr) {
+            channels[i] = c->get_output_pwm();
+        } else {
+            channels[i] = 1500; // Default to neutral if channel is null
```
"1500" - this is an argument as old as mode 2 vs mode 1 transmitters, and 1500 is *a* perfectly acceptable answer here. you could also have chosen 1520, but dont. :-)

peterbarker replied:
> Its a real flight controller on the other end. There is no right answer here, but I think 1500 is better than most.

These are normalised outputs, and they really are aetr.

Presumably `get_output_pwm` is getting scaled according to the `SERVOn_*`  parameters so all of those have to be right for the values to go into the CRSF AETR packet... that just seems a little awkward...

davidbuzz reviewed (approved):
re-reviewed, and flagged some of the now-resolved stuff.
I'm still for merging it, and i believe the majority of Peter's feedback has been taken onboard.

andyp1per committed:
autotest: nanny state

peterbarker commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,489 @@
+    scheduler.init(tasks, rate);
+    scheduler.set_task_rate(TaskIds::REPORTING, reporting_rate_hz);
+
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_CRSF_Out::crsf_out_thread, void), "crsf", 2048, AP_HAL::Scheduler::PRIORITY_RCOUT, 1)) {
```
Does this really require a 2kB stack?

andyp1per replied:
2k is pretty standard in the codebase - look at external AHRS. This is what I have tested with, I think I had a smaller one to start with and had to increase it

peterbarker commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,479 @@
+    }
+}
+
+void AP_CRSF_Out::update_rates_status()
```
Not being in your head, I don't understand the current name.

My reading of this was "OK, so this updates the rates' status".  But it does noes not do that.... I guess you mean it "updates the user about the current rates status"?

Perhaps `emit_rates_statustext`?

andyp1per replied:
I added a comment. This function gets a lot more functional in the next PR.

peterbarker commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,479 @@
+{
+    const float report_rate = _frontend._reporting_rate_hz.get();
+
+    GCS_SEND_TEXT(MAV_SEVERITY_INFO, "CRSFOut: RC: %uHz, Lat: %.1fms",
```
> Default is zero hz so no output, its also used for the logging rate in a future PR

You wouldn't have separate logging and reporting rates?

andyp1per replied:
at the moment no need and I really don't want to scope creep new features!

peterbarker commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,479 @@
+        return;
+    }
+
+    // find the configured CRSF output port
+    AP_HAL::UARTDriver* uart = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_CRSF_Output, 0);
+    if (uart == nullptr) {
+        return;
+    }
+
+    crsf_port = NEW_NOTHROW AP_RCProtocol_CRSF(AP::RC(), AP_RCProtocol_CRSF::PortMode::DIRECT_RCOUT, uart);
+
+    if (crsf_port == nullptr) {
+        debug_rcout("Init failed: could not create CRSF output port");
+        return;
+    }
```
Won't be extra stack (compilers are clever about re-use there.

feedback would be essentially unchanged.  Might actually be better, depending on initialisation order.

The only convincing one is the "don't want thread running" thing.  I kind of figured there was a reconfigure followed by a reboot in the user's future anyway so that didn't matter too much.  But it's a fair point.

peterbarker commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,489 @@
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_CRSF_Out::crsf_out_thread, void), "crsf", 2048, AP_HAL::Scheduler::PRIORITY_RCOUT, 1)) {
+        delete crsf_port;
+        crsf_port = nullptr;
+        debug_rcout("Failed to create CRSF_Out thread");
```
I'm wondering why you would ever really want to rumble on when this sort of thing fails.  If debug is disabled and the thread_create fails the user isn't told anything.  I would have thought an `allocation_failure` would be appropriate.

Similarly for the `crsf_port` allocation, actually.

andyp1per replied:
ok I have changed these two to GCS_SEND_TEXT

peterbarker commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,479 @@
+        }
+
+#ifdef CRSF_RCOUT_DEBUG
+        num_frames++;
+#endif
+        if (state == State::RUNNING && crsf_port->is_rx_active()) {
+            const bool send_rc_frame = pwm_is_fresh;
+            pwm_is_fresh = false;
+
+            if (send_rc_frame) {
+                scheduler.run_task_immediately(AETR_RC_FRAME);
+            }
+
+            if (!scheduler.update() && !send_rc_frame) {
+                send_heartbeat();
+            }
```
This one is still pending

andyp1per replied:
No it's not. I am asserting that it needs to be where it is and that I have tested it as such. These kind of structural changes that you are after stand a high risk of breaking things, I do not want to break things.

peterbarker replied:
There must be something I am not understanding here as it seems pretty clear to me this can go in the state machine in the same place you're *already* handling the state!

peterbarker reviewed (changes requested):
Sorry, another partial review only.  There was some confusion around how the outputs were configured which meant I misunderstood something fairly fundamental.

There are review comments from previous reviews still outstanding, too.

tridge commented on libraries/AP_CRSF/AP_CRSF_Protocol.cpp:
```diff
@@ -74,4 +105,293 @@ const char* AP_CRSF_Protocol::get_frame_type(uint8_t byte, uint8_t subtype)
+    uint8_t writeByteIndex = 1;
+    const uint8_t num_to_send = MIN(nchannels, MAX_CHANNELS);
+
+    for (uint8_t n = start_chan; n < num_to_send; n++) {
```
this loop is weird, if starting at start_chan then it won't send num_to_send unless start_chan is 0. Are the variables very badly named or does this actually miss things?

andyp1per replied:
I think you have found a bug. The idea is to only send 4 channels at high rate which works fine since they start at 0. Then the rest of the channels get sent but we are going to miss the final 4 channels if we calculate the exact number correctly. However in practice TBS firmware only supports 12 channels so we generally never hit this. But good catch!

timtuxworth replied:
ELRS uses CRSF and supports 16 channels.

andyp1per replied:
Fixed

andyp1per commented:
> There are review comments from previous reviews still outstanding, too.

Shouldn't be - also I am told that you do not expect every comment to be acted on, so it would be good to know what falls into this bucket.

andyp1per commented:
> Sorry, another partial review only

Just a reminder - you have already done a full review. It should not be necessary for you to do another full review as long as your concerns have been addressed. I am concious that both of our times are precious and we need to maximise the utility of both your input and my input. This feature is compiled out by default which in the past has been used to lower the bar to get complicated changes through which I would argue makes sense here.

andyp1per committed:
AP_CRSF_Protocol: encode the right number of channels in encode_variable_bit_channels()

andyp1per commented on libraries/AP_CRSF/AP_CRSF_Protocol.cpp:
```diff
@@ -0,0 +1,451 @@
+
+    for (uint8_t n = 0; n < numOfChannels; n++) {
+        while (bitsMerged < channelBits) {
+            if (readByteIndex >= CRSF_FRAME_PAYLOAD_MAX) {
```
actually this is now fixed with the other change you asked for.

andyp1per commented:
> ... come to think of it.... why are you going via servo outputs at all?! Why not have AP_MotorsOutput_CRSF_AETR? It would > be a better structure and more efficient. This file would either become AP_MotorsOutput_CRSF_AETR or this file would read > state from AP_MotorsOutput_CRSF_AETR or AP_MotorsOutput_CRSF_AETR would write state in here. But notionally you > could send the AETR packet as soon as the motor library updated its output.

We want all the channels not just the AETR values so going through the motor mixer doesn't make a lot of sense to me. Plus the current code works and is compiled out and nicely refactors some of the CRSF code. I don't think there is anything wrong with the current structure.

andyp1per committed:
AP_CRSF: make startup failure in CRSF out more explicit. Add comments to functions.

davidbuzz commented:
I've gone through all pending stuff that I could, and closed off stuff that is "fixed", "suggestion/wont-fix", "out of date" , etc.

peterbarker commented on libraries/AP_CRSF/AP_CRSF_Out.cpp:
```diff
@@ -0,0 +1,479 @@
+
+void AP_CRSF_Out::send_heartbeat()
+{
+    AP_CRSF_Protocol::Frame frame;
```
How much of this code is covered by the valgrind checks?

peterbarker commented on libraries/AP_CRSF/AP_CRSF_Protocol.cpp:
```diff
@@ -74,4 +105,300 @@ const char* AP_CRSF_Protocol::get_frame_type(uint8_t byte, uint8_t subtype)
+}
+
+// request for device info
+bool AP_CRSF_Protocol::process_device_info_frame(ParameterDeviceInfoFrame* info, VersionInfo* version, bool fakerx)
```
You're missing the nullptr check on `version`

Probably more places which I can find on a future review.

.... also, you're dereferencing-before-passing when encoding frames elsewhere in these patches.

peterbarker commented on libraries/AP_CRSF/AP_CRSF_Protocol.cpp:
```diff
@@ -0,0 +1,451 @@
+
+    const AP_FWVersion &fwver = AP::fwversion();
+    // write out the name with version, max width is 60 - 18 = the meaning of life
+    uint32_t n = strlen(fwver.fw_short_string);
+    strncpy((char*)info.payload, fwver.fw_short_string, 41);
+    n = MIN(n + 1, 42U);
```
I think you might be leaving `info.payload` with uninitialised stack data in it here.

peterbarker commented on libraries/AP_CRSF/AP_CRSF_Protocol.cpp:
```diff
@@ -74,4 +105,293 @@ const char* AP_CRSF_Protocol::get_frame_type(uint8_t byte, uint8_t subtype)
+    strncpy((char*)info.payload, fwver.fw_short_string, 41);
+    n = MIN(n + 1, 42U);
+
+    put_be32_ptr(&info.payload[n], fwver.os_sw_version);   // serial number
+    n += 4;
```
This makes me very itchy.

The sort of code that can walk off the end of an array when someone screws up.

This could just as easily have been cast-to-struct and assignment.

peterbarker commented on libraries/AP_RCProtocol/AP_RCProtocol.cpp:
```diff
@@ -401,13 +415,17 @@ void AP_RCProtocol::check_added_uart(void)
+// called at 10Hz from AP_Vehicle::update_rc()
 void AP_RCProtocol::update()
 {
+#if !APM_BUILD_TYPE(APM_BUILD_UNKNOWN)
```
Why is this actually required?

I mean `AP_RCPROTOCOL_ADDED_UART_ENABLED` would seem to be appropriate.  But setting that based on `APM_BUILD_TYPE` could be problematic, thus me asking when we would want to set it.

peterbarker commented on libraries/AP_RCProtocol/AP_RCProtocol.h:
```diff
@@ -27,6 +27,10 @@
 #define MAX_RCIN_CHANNELS 18
 #define MIN_RCIN_CHANNELS  5
 
+#ifndef RCPROTOCOL_DEBUG_DETECTION
```
```suggestion
#ifndef AP_RCPROTOCOL_DEBUG_DETECTION
```

(but I won't insist on it... it's just convention)

peterbarker commented on Tools/autotest/arducopter.py:
```diff
@@ -13289,8 +13289,146 @@ def PerfInfo(self):
+
+            # On the RX side, verify we're receiving RC channels
+            self.progress("Checking RX vehicle receives RC channels")
+            # Put vehicle in stabilize so that there are actually some outputs on channel 1-4
```
```suggestion
            # Put vehicle in a manual-throttle-mode so that there are actually some outputs on channel 1-4
```

lthall reviewed (commented):
I personally don't think we should be allowing this support into ArduPilot at all.

Why are we trying to add and carry support for other flight controllers!

I am happy for this to live in someones other branch but not in the main code to be supported for all time.

davidbuzz commented:
> I personally don't think we should be allowing this support into ArduPilot at all.
> 
> Why are we trying to add and carry support for other flight controllers!
> 
> I am happy for this to live in someones other branch but not in the main code to be supported for all time.

There is a number of precendents for dual-flight-controller setups, eg the Canberra Trad Heli with a rate-controller, eg  the use of a dual-inline flight controllers for FTS capabilities in the Outback challenge ( i personaly ran dual-controllers there ),  not to mention redundancy systems that B*eing did that one time with dual flight controlers,   etc.   I'm sure there's more.     This particular one is about using the "rate" part of betaflight as the down-stream system, but its also about leveraging the fact that a well-tuned betaflight can be very high performance, AND you dont have , necessarily, a single pint of failure any more.

tridge commented:
I think the capability is in principle a good thing - separate from the implementation details, I can see the example of RPiZero + external rate controller as quite a nice thing to be able to support

lthall commented:
> AND you dont have , necessarily, a single pint of failure any more.

What ? How is that an argument here? How is it an argument at all!

As for the rest of your statement...... We are doing this because it could be called a "dual-flight-controller setup" and others have done it sooooooo that makes this a good idea....

The only reason to do this is because the target for this can't run the full stack on their own. So we are doing this to add support for a target that out of the box can't run ArduPilot. Where is the crazy eyed emoji? (found it)

On top of that we now need to be responsible for ensuring BetaFlight stays compatible with this.

It must be April fools or something.

lthall commented:
> I think the capability is in principle a good thing - separate from the implementation details, I can see the example of RPiZero + external rate controller as quite a nice thing to be able to support

Why. Just use a RPiZero + an external autopilot that can run the full autopilot firmware. Literally the same size. In fact the rate controller being used is probably more capable of running autopilot and achieving better overall performance.

The only reason I can think of doing this is "kindly sponsored by Raspberry Pi". And that doesn't benefit the project, it costs the project.

timtuxworth commented:
> The only reason to do this is because the target for this can't run the full stack on their own. So we are doing this to add support for a target that out of the box can't run ArduPilot. Where is the crazy eyed emoji? (found it)

Not at all @lthall - one example could be 2 or 3 different AutoPilots running different platforms. 

I have done something similar with 2 AutoPilots running Plane, but I can imagine having an AP running Rover connected to an AP running Plane or Copter, for example as a way to handing taxi-ing on the ground (rover) and flying (plane/copter). I'm sure there could be other use cases once we have the capability.

lthall commented:
> Not at all @lthall - one example could be 2 or 3 different AutoPilots running different platforms.
> 
> I have done something similar with 2 AutoPilots running Plane, but I can imagine having an AP running Rover connected to an AP running Plane or Copter, for example as a way to handing taxi-ing on the ground (rover) and flying (plane/copter). I'm sure there could be other use cases once we have the capability.

What does that have to do with this PR. This is the same argument Buzz made. Multiple redundant autopilots. This isn't multiple redundant autopilots. Why are we talking about multiple autopilots?

My understanding of this is Andy was payed to make help Raspberry Pi build an autopilot using the PiZero2W. But it isn't capable of supporting the whole stack so the solution was to connect it to another autopilot to do the rate control and pass the gyro data back to the PiZero2W so it could run autopilot back to the attitude controller and not have to do all the high rate stuff. So why not set it up as a companion computer and use ArduCopter on the second flight controller? Because that means it isn't an autopilot and someone in Raspberry pi sold it to management as a way to get into this huge killer drone market that everybody is talking about (speculation).

This isn't two redundant autopilots. It isn't two autopilots connected together. This isn't support for Arducopter split over two processors. 

And as you should expect. The performance is crap. If you want to get 12" performance out of a 5" quad this is your solution (and I am not talking about the endurance). 

This isn't a passion project for Andy, it is a payed project that he wants to complete so he can move on to the next payed project. That means what ever we let into the code is what people like @peterbarker will be asked to fix or work around FOR EVER. And we will pay for it in slower PR reviews and less code clean up.

So if this is a new communication protocol then sure. Lets do it a new communication protocol. Lets make sure it is implemented to be as future proof as possible. Lets make it an addition worth supporting.

If it is split processor support that is so attractive then lets do that. Design it, Plan it, and do a good job. That is not what is happening here!

rmackay9 commented:
I think one important issue with the PR is that it costs about 1500 bytes of flash even when disabled.

rmackay9 commented:
Considering the objections to the implementation, the flash cost and the experimental nature of the PR, I think the best approach would be for AndyP to maintain this in a separate branch (either within his own ardupilot repo or within the main ArduPilot repo).  I believe there are even ways to expose a private repo to the custom build server meaning users could build firmware for their own board.

This approach would allow AndyP to bear the ongoing support costs while we determine if there's any community interest in the feature.  We could come back and revisit in 6 months or a year and see if there's been any uptake.

We can discuss at the next dev call but perhaps @IamPete1 should weigh in as well.

davidbuzz commented:
id just like to point out that a lot of statements made by leonard on this PR are misinformed.

andyp1per commented:
I’d like to address the concerns raised by @lthall and @rmackay9. While I understand the desire to protect the codebase from "niche" features, I believe there is a fundamental misunderstanding of the architectural value this PR provides.

Leonard, you mentioned this isn't "ArduPilot split over two processors." Actually, that is exactly the door this opens. This implementation follows the Heterogeneous Multi-Processing pattern used in professional aerospace. By moving high-rate, real-time tasks (Rate Control/IMU) to a dedicated MCU while leaving high-level logic (EKF/Nav) on a Linux SoC, we create a "Safety Core" architecture. This is functionally analogous to how our IOMCU works for fixed-wing, providing a deterministic fallback that Copter currently lacks on single-processor Linux builds.

Regarding the "12-inch performance" comment: Leonard is correct that a round-trip over CRSF introduces more latency than an internal bus. We are not claiming this is for a world-class 5" racing drone. However, at 2Mbit, the latency is well-managed.
In flight testing, the aircraft performed significantly better than the vast majority of commercial tunes I encounter. For the target use case—bringing ArduPilot's navigation to high-compute Linux hardware—this is a deliberate and successful engineering trade-off. The flight logs show it is a stable, reliable, and highly capable platform.

As the primary maintainer of CRSF, I am not "completing a project and moving on." I am evolving the subsystem I am responsible for to support where the hardware market is heading.

- This code is isolated within the CRSF protocol layer.
- It is compiled out by default.
- The footprint is a mere 260 bytes.
- It gives me a much better path for ongoing, solid testing of the CRSF stack itself.

If the domain maintainer sees a path to better testing and modularity with zero impact on the core "fast loop" for other users, that should carry significant weight.

We have a long-standing history of accepting specialized hardware support when it is well-isolated and backed by a maintainer (e.g., QURT). This PR follows that pattern but with even lower risk. Blocking maintainer-backed innovation because it doesn't fit the traditional monolithic MCU model risks leaving ArduPilot behind as the industry moves toward distributed, multi-compute architectures.

I appreciate the detail in Peter's cleanup suggestions, but we must be careful not to let "perfect" become the enemy of "verifiable." Mixing extensive structural refactoring with complex protocol logic makes auditing the behavioral impact much harder.

On the subjective points regarding code structure, I’d like to exercise the author’s casting vote to keep the current structure. It is what I have validated in flight, and I believe it keeps the protocol logic clearest. Let’s focus the remaining review on correctness and safety so we can move this forward.

lthall commented:
> This implementation follows the Heterogeneous Multi-Processing pattern used in professional aerospace. By moving high-rate, real-time tasks (Rate Control/IMU) to a dedicated MCU while leaving high-level logic (EKF/Nav) on a Linux SoC, we create a "Safety Core" architecture. This is functionally analogous to how our IOMCU works for fixed-wing, providing a deterministic fallback that Copter currently lacks on single-processor Linux builds.

> Blocking maintainer-backed innovation because it doesn't fit the traditional monolithic MCU model risks leaving ArduPilot behind as the industry moves toward distributed, multi-compute architectures.

I don't think you are accurately representing the direction of the "industry" here.

Splitting the control loops right in the middle is only a solution if you don't have a choice. It is not the future of any professional aerospace industry. This is not a safety core. This is just an attempt to turn Autopilot into a Beta-flight accessory (and make us pay for it).

>> In flight testing, the aircraft performed significantly better than the vast majority of commercial tunes I encounter.

Exactly, this is an extremely low bar for a 5 inch quad.

>> For the target use case—bringing ArduPilot's navigation to high-compute Linux hardware—this is a deliberate and successful engineering trade-off. The flight logs show it is a stable, reliable, and highly capable platform.

Sure. I am just saying that this isn't a trade off that is in ArduPilots best interested.

>> I am evolving the subsystem I am responsible for to support where the hardware market is heading.

I do not agree that this is where the hardware market is heading. Sure this might be one option that vendors looking to get into weaponized multirotors may be exploring but that is not the hardware market that this project supports.

The only justification I can think of for this particular mix of engineering decisions is a disposable, light weight plug in co-processor to provide basic navigation to existing beta-flight aircraft. Maybe using optical flow or terrain imagery. Looking at the rest of your work it seems to point heavily in that direction....

> I appreciate the detail in Peter's cleanup suggestions, but we must be careful not to let "perfect" become the enemy of "verifiable." Mixing extensive structural refactoring with complex protocol logic makes auditing the behavioral impact much harder.

I think what you are really saying here is you don't want to do the work to get the PR up to a standard that other developers are happy to accept it.

> On the subjective points regarding code structure, I’d like to exercise the author’s casting vote to keep the current structure. It is what I have validated in flight, and I believe it keeps the protocol logic clearest. Let’s focus the remaining review on correctness and safety so we can move this forward.

I think this approach is best suited to maintaining your own branch while you develop it as you see fit. That is where the "author's casting vote" has weight.

proficnc commented:
This brings the framework of support for Ardupilot and a rate control autopilot to run on the dual core cube orange plus. 

This is one of multiple methods being developed for this purpose and is extremely useful. 

Thanks for your efforts in pushing this forward @andyp1per

lthall commented:
> This brings the framework of support for Ardupilot and a rate control autopilot to run on the dual core cube orange plus.
> 
> This is one of multiple methods being developed for this purpose and is extremely useful.
> 
> Thanks for your efforts in pushing this forward @andyp1per

Ok, that makes sense. So you are saying HEX wants to run BetaFlight on one core and ArduCopter on the other? I guess HEX also wants ArduPilot to be just a BetaFlight flight mode as well?

davidbuzz commented:
dev team who have publically said they support this: Tridge, Randy ( verbally last dev call ) , me, Andy, Phillip (above).   Its also useful in ardu+ardu configurations like the dual-core-Cube, and almost certainly other dual-cpu setups.

lthall commented:
> dev team who have publically said they support this: Tridge, Randy ( verbally last dev call ) , me, Andy, Phillip (above). Its also useful in ardu+ardu configurations like the dual-core-Cube, and almost certainly other dual-cpu setups.

@davidbuzz Well I am glad you are here to tell us what they think. It is such a shame they can't speak for themselves.
